### PR TITLE
Sync from internal for 2 12

### PR DESCRIPTION
**Pull request by aws-rhsoln**

### CODE REVIEW

**Review by jeffhataws — APPROVED** [review]

File: src/neuronx_distributed/parallel_layers/mappings.py [hunk 1]
```diff
@@ -141,6 +198,46 @@ def backward(ctx, grad_output):
...
+        if ctx.to_model_parallel:
+            return _reduce_scatter_along_first_dim(grad_output), None
+        else:
+            return _split_along_first_dim(grad_output), None
+
+
+class _ReduceScatterToSequenceParallelRegion(torch.autograd.Function):
+    """Reduce scatter the input from the sequence parallel region and concatenate."""
+
+    # FIXME(mkozuki): Definition of static symbolic methods don't look correct according to
```
Comment (aws-sadaf) [inline src/neuronx_distributed/parallel_layers/mappings.py:226]: Do we want to keep these comments in this external repo?

File: src/neuronx_distributed/parallel_layers/loss_functions.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        # All the inputs have softmax as thier gradient.
+        grad_input = softmax
+        # For simplicity, work with the 2D gradient.
+        partition_vocab_size = softmax.size()[-1]
+        grad_2d = grad_input.view(-1, partition_vocab_size)
+
+        # Add the gradient from matching classes.
+        arange_1d = torch.arange(start=0, end=grad_2d.size()[0], device=grad_2d.device, dtype=torch.long)
+
+        # softmax_update = 1.0 - target_mask.view(-1).float()
```
Comment (aws-sadaf) [inline src/neuronx_distributed/parallel_layers/loss_functions.py:107]: Same for this one?

File: src/neuronx_distributed/parallel_layers/mappings.py [hunk 3]
```diff
@@ -141,6 +198,46 @@ def backward(ctx, grad_output):
         return _split_along_last_dim(grad_output)
 
 
+class _GatherFromSequenceParallelRegion(torch.autograd.Function):
+    """Gather the input from sequence parallel region and concatenate."""
+
+    # FIXME(mkozuki): Definition of static symbolic methods don't look correct according to
+    # https://pytorch.org/docs/stable/onnx.html#static-symbolic-method
```
Comment (aws-sadaf) [inline src/neuronx_distributed/parallel_layers/mappings.py:205]: Here too?

File: src/neuronx_distributed/parallel_layers/mappings.py (hunk 1 above)
Comment (aws-rhsoln) [reply src/neuronx_distributed/parallel_layers/mappings.py:226]: These are coming from the apex library, its okay to have those

File: src/neuronx_distributed/parallel_layers/loss_functions.py (hunk 2 above)
Comment (aws-rhsoln) [reply src/neuronx_distributed/parallel_layers/loss_functions.py:107]: Okay to have those

**Review by aws-sadaf — APPROVED** [review]

**Pull request merged**